extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/dart/controller/http/DartSqlResource.java

```diff
@@ -0,0 +1,264 @@
+      response = new GetQueriesResponse(
+          queries.stream()
+                 .filter(
+                     query ->
+                         authenticationResult.getAuthenticatedBy() != null
+                         && authenticationResult.getIdentity() != null
+                         && Objects.equals(authenticationResult.getAuthenticatedBy(), query.getAuthenticator())
+                         && Objects.equals(authenticationResult.getIdentity(), query.getIdentity()))
+                 .map(DartQueryInfo::withoutAuthenticationResult)
+                 .collect(Collectors.toList())
+      );
+    }
+
+    AuthorizationUtils.setRequestAuthorizationAttributeIfNeeded(req);
+    return response;
+  }
+
+  /**
+   * API to issue a query.
+   */
+  @POST
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Override
+  public Response doPost(
+      final SqlQuery sqlQuery,
+      @Context final HttpServletRequest req
+  )
+  {
+    final Map<String, Object> context = new HashMap<>(sqlQuery.getContext());
+
+    // Default context keys from dartQueryConfig.
+    for (Map.Entry<String, Object> entry : dartQueryConfig.getContext().entrySet()) {
+      context.putIfAbsent(entry.getKey(), entry.getValue());
+    }
+
+    // Dart queryId must be globally unique; cannot use user-provided sqlQueryId or queryId.
+    final String dartQueryId = UUID.randomUUID().toString();
+    context.put(DartSqlEngine.CTX_DART_QUERY_ID, dartQueryId);
+
+    return super.doPost(sqlQuery.withOverridenContext(context), req);
+  }
+
+  /**
+   * API to cancel a query.
+   */
+  @DELETE
+  @Path("{id}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Override
+  public Response cancelQuery(
+      @PathParam("id") String sqlQueryId,
+      @Context final HttpServletRequest req
+  )
+  {
+    log.debug("Received cancel request for query[%s]", sqlQueryId);
+
+    List<SqlLifecycleManager.Cancelable> cancelables = sqlLifecycleManager.getAll(sqlQueryId);
+    if (cancelables.isEmpty()) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+
+    final Access access = authorizeCancellation(req, cancelables);
+
+    if (access.isAllowed()) {
+      sqlLifecycleManager.removeAll(sqlQueryId, cancelables);
+
+      // Don't call cancel() on the cancelables. That just cancels native queries, which is useless here. Instead,
+      // get the controller and stop it.
+      boolean found = false;
+      for (SqlLifecycleManager.Cancelable cancelable : cancelables) {
+        final HttpStatement stmt = (HttpStatement) cancelable;
+        final Object dartQueryId = stmt.context().get(DartSqlEngine.CTX_DART_QUERY_ID);
+        if (dartQueryId instanceof String) {
+          final ControllerHolder holder = controllerRegistry.get((String) dartQueryId);
+          if (holder != null) {
+            found = true;
+            holder.cancel();
+          }
+        }
```

Comment:
> can we log when the dartQueryId is not of a string type or is null

Yes, added.

> It would be nice for this logic to sit inside the cancellable itself but that would require implementing a DartStatement I suppose.

Yeah it would be a new type of statement. But the way the statement stuff is designed, it isn't easily extensible, so it would be a bigger project to add new statement types. Even the `SqlStatementResource` (MSQ async query resource) uses `HttpStatement`.